### PR TITLE
Allow solution pack to accept JP2 files natively

### PR DESCRIPTION
**Pull request by dltj**
Replaces https://github.com/Islandora/islandora_solution_pack_large_image/pull/85

Also removes the module.install requirement that the kdu-compress executable be on the Apache PATH. We're specifying the path as part of the configuration, so there is no reason why the executable should be available at module install time.

This module makes use of kdu-expand to create TIFF images from JP2 OBJ datastreams (e.g. where there was no original TIFF ingested but a TIFF derivative is needed), so the 'islandora_kakadu_url' variable was changed to point to the directory where the Kakadu executables live. An update hook is provided to modify the variable.

Based on work by @grf (Randy Fischer of FLVC).

### CODE REVIEW

File: includes/derivatives.inc [hunk 1]
```diff
@@ -50,10 +50,19 @@ function islandora_large_image_create_jp2_derivative(AbstractObject $object, $fo
     $kakadu = variable_get('islandora_use_kakadu', !islandora_large_image_check_imagemagick_for_jpeg2000());
     $mime = strtolower($object['OBJ']->mimetype);
     if ($mime == 'image/jpeg' || $mime == 'image/jpg') {
+      // Force jpeg conversions through ImageMagick rather than Kakadu.
```
Comment (dltj) [inline includes/derivatives.inc:53]: I would note that in this new pattern of derivative creation, the return value of $uploaded_file (7 lines above this one) is not tested for a FALSE value (presumably if the OBJ datastream didn't exist for some reason) before trying to use it.  The previous pattern did (https://github.com/Islandora/islandora_solution_pack_large_image/blob/7.x/includes/derivatives.inc#L482).

**Comment by dltj** [comment]
This should work in theory.  The post-7.x-1.2-created pattern for creating derivatives actually cleaned up the code quite a bit (in spite of my ranting on the IRC channel).  Unfortunately, I'm not in a position to take my Islandora instance all the way to HEAD (from 7.x-1.2, where it is now) to try it out.  Can someone try this out, please?  If it fails, let it sit until I update my environment to the next release (when I'm more in sync with what is at HEAD now).

File: includes/derivatives.inc (hunk 1 above)
Comment (jordandukart) [inline includes/derivatives.inc:53]: That function exists for legacy purposes, that isn't actually used in derivative creation. If you look at https://github.com/Islandora/islandora_solution_pack_large_image/blob/7.x/includes/derivatives.inc#L59-L68 you can see that this is where this code is.

Comment (dltj) [inline includes/derivatives.inc:53]: Thanks for the explanation in the IRC channel about how there must be an OBJ datastream in the object before the derivative-creation functions will be called.

Comment (grf) [inline includes/derivatives.inc:53]: Peter

thanks for keeping me in the loop on this.  You may be aware I've been on
medical leave for the last 4 months but am resuming work soon.

Best, and happy holidays.

-Randy

File: includes/derivatives.inc [hunk 2]
```diff
@@ -50,10 +50,19 @@ function islandora_large_image_create_jp2_derivative(AbstractObject $object, $fo
...
       $kakadu = FALSE;
     }
 
-    if ($kakadu) {
+    // If we already have a JPEG2000 file, just copy it to the JP2 datastream.
```
Comment (adam-vessey) [inline includes/derivatives.inc:57]: Source JP2s could be lossless... not sure it's such a good idea to serve 'em up for viewing, as they could be of significant size.

File: includes/derivatives.inc [hunk 3]
```diff
@@ -255,7 +264,13 @@ function islandora_large_image_kdu_compress($src, $dest, $args = NULL) {
   $ret = -1;
   $ret_palette = 0;
   $output = array();
-  $kdu_compress = variable_get('islandora_kakadu_url', 'kdu_compress');
+  $kdu_home = variable_get('islandora_kakadu_url', FALSE);
```
Comment (willtp87) [inline includes/derivatives.inc:267]: Changing the default value of a Drupal variable is API breaking.  A new variable should be considered.

File: includes/derivatives.inc [hunk 4]
```diff
@@ -281,6 +296,51 @@ function islandora_large_image_kdu_compress($src, $dest, $args = NULL) {
...
+ * @param string $dest
+ *   The destination file which kdu_expand will generate.
+ * @param string $args
+ *   Any additional arguments for kdu_expand, if not provided reasonable
+ *   defaults will be used.
+ *
+ * @return bool
+ *   Returns the newly generated file uri or FALSE if the conversion failed
+ */
+function islandora_large_image_kdu_expand($src, $dest, $args = NULL) {
```
Comment (willtp87) [inline includes/derivatives.inc:312]: Is this used?

File: includes/derivatives.inc [hunk 5]
```diff
@@ -255,7 +264,13 @@ function islandora_large_image_kdu_compress($src, $dest, $args = NULL) {
   $ret = -1;
   $ret_palette = 0;
   $output = array();
-  $kdu_compress = variable_get('islandora_kakadu_url', 'kdu_compress');
+  $kdu_home = variable_get('islandora_kakadu_url', FALSE);
+  if ($kdu_home) {
+    $kdu_compress = "$kdu_home/kdu_compress";
```
Comment (willtp87) [inline includes/derivatives.inc:269]: The assumption regarding the executable name is a regression.

**Comment by willtp87** [comment]
A long time with no activity, should this be closed? is it going to be picked up again?

**Comment by dltj** [comment]
I want to pick this back up again, but our internal development is not yet up to HEAD.  Thanks @adam-vessey and @willtp87 for the great comments.

**Comment by dltj** [comment]
Yup -- that'd be fine.  Who can I thank for adding it to HEAD?

**Comment by adam-vessey** [comment]
Errr... Only looks to "support" tiffs:
- https://github.com/Islandora/islandora_solution_pack_large_image/blob/7.x/includes/image_upload.form.inc#L21
- https://github.com/Islandora/islandora_solution_pack_large_image/blob/7.x/xml/islandora_large_image_ds_composite_model.xml#L27

... Though really, there not much else jumping to the front of my mind which should be necessary to make it work...?

**Comment by adam-vessey** [comment]
Yeah, kinda what I'm getting at... Pretty sure the underlying code can handle them... Just not sure the UI will allow them as it is, currently.

**Comment by adam-vessey** [comment]
I'm inclined to say "yes", but at the same time, it feels weird to have a feature be implemented with out being directly testable? Be able to handle them being part of this module, but only being usable externally seems... Weird? Dunno...  Unless part of "being able to handle it" comprises throwing together a (couple of?) test(s) handling additional input?

**Comment by dltj** [comment]
@ruebot -- Up above you ask

> One last confirmation, the intention with this and Islandora/islandora_ocr#31 pull request were to ingest JP2s, and have OCR happen? If so, I did this in the development VM I was working with today, and that is what all of the books here are. If that is the case, I'll create a Jira improvement ticket too... if one doesn't exist already.

Yes that was the intention.  As I recall, Tesseract takes only TIFFs, so one had to convert the JP2 to TIFF before feeding it into the OCR engine.

**Pull request closed without merging**